### PR TITLE
planner: fix CTE bug when MergeJoin is used

### LINKED ISSUE #25474: Got error when use merge_join hint in CTE
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table if exists t1, t2;
    create table t1(c1 int);
    insert into t1 values(1), (2), (1), (2);
    create table t2(c1 int primary key);
    insert into t2 values(1), (2), (3);
    with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from t1 where c1 < 3) select /*+ MERGE_JOIN(dt1, dt2) */ * from cte1 dt1 left join t1 dt2 on dt1.c1 = dt2.c1 order by dt1.c1, dt2.c1;

### 2. What did you expect to see? (Required)

    mysql>     with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from t1 where c1 < 3) select /*+ MERGE_JOIN(dt1, dt2) */ * from cte1 dt1 left join t1 dt2 on dt1.c1 = dt2.c1 order by dt1.c1, dt2.c1;
    +------+------+
    | c1   | c1   |
    +------+------+
    |    1 |    1 |
    |    1 |    1 |
    |    2 |    2 |
    |    2 |    2 |
    |    3 | NULL |
    +------+------+
    5 rows in set (0.00 sec)
### 3. What did you see instead (Required)

    mysql>     with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from t1 where c1 < 3) select /*+ MERGE_JOIN(dt1, dt2) */ * from cte1 dt1 left join t1 dt2 on dt1.c1 = dt2.c1 order by dt1.c1, dt2.c1;
    ERROR 1815 (HY000): Internal : Can't find a proper physical plan for this query
### 4. What is your TiDB version? (Required)

master and 5.1

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #25474 

Problem Summary: When use MERGE JOIN to join a CTE table, planner give errors

    mysql>     with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from t1 where c1 < 3) select /*+ MERGE_JOIN(dt1, dt2) */ * from cte1 dt1 left join t1 dt2 on dt1.c1 = dt2.c1 order by dt1.c1, dt2.c1;
    ERROR 1815 (HY000): Internal : Can't find a proper physical plan for this query

### What is changed and how it works?

What's Changed: Change logic in `LogicalCTE.findBestTask`, when prop requires sort item, we add a sort operator explicitly.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects: no

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- planner: fix CTE bug when MergeJoin is used

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -1962,30 +1962,32 @@ func (ds *DataSource) getOriginalPhysicalIndexScan(prop *property.PhysicalProper
...
+				return nil, 1, err
+			}
+		}
+
+		pcte = PhysicalCTE{SeedPlan: sp, RecurPlan: rp, CTE: p.cte, cteAsName: p.cteAsName}.Init(p.ctx, p.stats)
+		pcte.SetSchema(p.schema)
+		p.cte.physicalCTE = pcte
+	}
+	t = &rootTask{pcte, pcte.SeedPlan.statsInfo().RowCount, false}
+	if !prop.IsEmpty() {
```
Comment (wjhuang2016) [inline planner/core/find_best_task.go:1988]:
``` sql
mysql> desc with recursive cte1(c1) as (select c1 from t1 order by c1) select /*+ MERGE_JOIN(dt1, dt2) */ * from cte1 dt1 left join t1 dt2 on dt1.c1 = dt2.c1;
+--------------------------------+---------+-----------+---------------+------------------------------------------------------------+
| id                             | estRows | task      | access object | operator info                                              |
+--------------------------------+---------+-----------+---------------+------------------------------------------------------------+
| MergeJoin_9                    | 1.25    | root      |               | left outer join, left key:test.t1.c1, right key:test.t1.c1 |
| ├─Sort_21(Build)               | 4.00    | root      |               | test.t1.c1                                                 |
| │ └─TableReader_20             | 4.00    | root      |               | data:Selection_19                                          |
| │   └─Selection_19             | 4.00    | cop[tikv] |               | not(isnull(test.t1.c1))                                    |
| │     └─TableFullScan_18       | 4.00    | cop[tikv] | table:dt2     | keep order:false, stats:pseudo                             |
| └─Sort_17(Probe)               | 1.00    | root      |               | test.t1.c1                                                 |
|   └─CTEFullScan_16             | 1.00    | root      | CTE:dt1       | data:CTE_0                                                 |
| CTE_0                          | 1.00    | root      |               | Non-Recursive CTE                                          |
| └─Sort_10(Seed Part)           | 4.00    | root      |               | test.t1.c1                                                 |
|   └─TableReader_14             | 4.00    | root      |               | data:TableFullScan_13                                      |
|     └─TableFullScan_13         | 4.00    | cop[tikv] | table:t1      | keep order:false, stats:pseudo                             |
+--------------------------------+---------+-----------+---------------+------------------------------------------------------------+
11 rows in set (0.00 sec)
```

**Review by wjhuang2016 — APPROVED** [review]

File: planner/core/find_best_task.go (hunk 1 above)
Comment (wjhuang2016) [reply planner/core/find_best_task.go:1991]: Please add this test case.

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -1962,30 +1962,37 @@ func (ds *DataSource) getOriginalPhysicalIndexScan(prop *property.PhysicalProper
...
+		// When got here, `prop.CanAddEnforcer` must be true,
+		// so we just enforce property.
+		t = enforceProperty(prop, t, pcte.basePlan.ctx)
```
Comment (eurekaka) [inline planner/core/find_best_task.go:1994]: We may fail to find the optimal plan for `with recursive cte1(c1) as (select c1 from t1) select /*+ MERGE_JOIN(dt1, dt2) */ * from cte1 dt1 join t2 dt2 on dt1.c1 = dt2.c1` if `t1` has an index on `c1`, i.e, for the cte part, using index on `c1` may be better than enforcing a `Sort` here, but we are unaware of this sort property when calling `DoOptimize` for the cte definitions.

**Review by eurekaka — APPROVED** [review]

**Comment by eurekaka** [comment]
We can mark https://github.com/pingcap/tidb/pull/25514#discussion_r656064569 as an enhancement issue to be resolved later, so I approve this PR.

File: planner/core/find_best_task.go (hunk 2 above)
Comment (guo-shaoge) [reply planner/core/find_best_task.go:1994]: For now, the optimization of `seed part` is seperated. So I record this problem in another issue(#25674)

File: planner/core/find_best_task.go (hunk 1 above)
Comment (guo-shaoge) [reply planner/core/find_best_task.go:1991]: Done. This is not the optimial plan. Record in this [issue](https://github.com/pingcap/tidb/issues/25674)

Comment (guo-shaoge) [reply planner/core/find_best_task.go]: To fix this problem, following changes are done:
1. Change the declaration of `physicalOptimize()`, add `prop` argument.
2. Add `doOptimize()` function, add `prop` argument.
3. Add `(p *LogicalCTE) replaceSortItems` function. This function will replace columns in prop.SortItems.
4. Add `seedColMap` in `CTEClass`. Record mapping of `UniqueID` of column in CTE output schema and `original column` of seed part.

File: cmd/explaintest/r/cte.result [hunk 3]
```diff
@@ -170,3 +379,220 @@ CTE_0	1.00	root		Non-Recursive CTE
...
+3	1
+3	2
+// No Sort on CTE. Because we have order by in CTE definition.
+explain with cte1 as (select c1 from t1 order by c1) select /*+ MERGE_JOIN(dt1, dt2) */ * from t1 dt1 inner join cte1 dt2 on dt2.c1 = dt1.c1 order by 1, 2;
+id	estRows	task	access object	operator info
+MergeJoin_30	1.00	root		inner join, left key:test.t1.c1, right key:test.t1.c1
+├─Sort_28(Build)	0.80	root		test.t1.c1
+│ └─Selection_20	0.80	root		not(isnull(test.t1.c1))
+│   └─CTEFullScan_27	1.00	root	CTE:dt2	data:CTE_0
+└─Sort_18(Probe)	9990.00	root		test.t1.c1
```
Comment (guo-shaoge) [inline cmd/explaintest/r/cte.result:438]: Maybe an extra sort, need fix.

**Comment by guo-shaoge** [comment]
@wjhuang2016 ptal~

File: cmd/explaintest/r/cte.result [hunk 4]
```diff
@@ -170,3 +379,205 @@ CTE_0	10000.00	root		Non-Recursive CTE
...
+2	2
+2	2
+2	2
+2	2
+// No Sort on CTE. Because tpk.c1 is pk.
+explain with cte1 as (select c1 from tpk) select /*+ MERGE_JOIN(dt1, dt2) */ * from tpk dt1 inner join cte1 dt2 on dt2.c1 = dt1.c1 order by 1, 2;
+id	estRows	task	access object	operator info
+MergeJoin_24	10000.00	root		inner join, left key:test.tpk.c1, right key:test.tpk.c1
+├─Sort_22(Build)	8000.00	root		test.tpk.c1
+│ └─Selection_20	8000.00	root		not(isnull(test.tpk.c1))
```
Comment (guo-shaoge) [inline cmd/explaintest/r/cte.result:416]: An extra sort operator, because tpk.c1 is orded.

**Comment by guo-shaoge** [comment]
@wjhuang2016 ptal

File: planner/core/find_best_task.go [hunk 5]
```diff
@@ -1972,12 +1972,21 @@ func (ds *DataSource) getOriginalPhysicalIndexScan(prop *property.PhysicalProper
...
+	if !prop.IsEmpty() && !prop.CanAddEnforcer {
 		return invalidTask, 1, nil
 	}
-	if p.cte.cteTask != nil {
+	t = p.cte.cteTask
+	defer func() {
+		if err != nil {
+			return
+		}
+		if !prop.IsEmpty() && prop.CanAddEnforcer {
```
Comment (wjhuang2016) [inline planner/core/find_best_task.go:1983]: prop.CanAddEnforcer must be true here.

File: planner/core/find_best_task.go [hunk 6]
```diff
@@ -1972,12 +1972,21 @@ func (ds *DataSource) getOriginalPhysicalIndexScan(prop *property.PhysicalProper
 }
 
 func (p *LogicalCTE) findBestTask(prop *property.PhysicalProperty, planCounter *PlanCounterTp) (t task, cntPlan int64, err error) {
-	if !prop.IsEmpty() {
+	if !prop.IsEmpty() && !prop.CanAddEnforcer {
 		return invalidTask, 1, nil
 	}
-	if p.cte.cteTask != nil {
+	t = p.cte.cteTask
+	defer func() {
```
Comment (wjhuang2016) [inline planner/core/find_best_task.go:1979]: We better add some tests for the case that different prop on CTE

**Comment by guo-shaoge** [comment]
@wjhuang2016 ptal

File: cmd/explaintest/r/cte.result [hunk 7]
```diff
@@ -170,3 +379,231 @@ CTE_0	10000.00	root		Non-Recursive CTE
...
+  └─TableFullScan_17	10000.00	cop[tikv]	table:dt1	keep order:true, stats:pseudo
+CTE_0	10000.00	root		Non-Recursive CTE
+└─TableReader_12(Seed Part)	10000.00	root		data:TableFullScan_11
+  └─TableFullScan_11	10000.00	cop[tikv]	table:tpk	keep order:false, stats:pseudo
+with cte1 as (select c1 from tpk) select /*+ MERGE_JOIN(dt1, dt2) */ * from tpk dt1 inner join cte1 dt2 on dt2.c1 = dt1.c1 order by 1, 2;
+c1	c1
+1	1
+2	2
+3	3
+// No Sort on CTE. Because we have order by in CTE definition.
```
Comment (wjhuang2016) [inline cmd/explaintest/r/cte.result:428]: Need fix

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**